core/shared/src/main/scala/spire/math/Polynomial.scala

```diff
@@ -285,6 +280,59 @@ trait Polynomial[@sp(Double) C] { lhs =>
   }
 
   /**
+   * Shift this polynomial along the x-axis by `h`, so that `this(x + h) ==
+   * this.shift(h).apply(x)`.  This is equivalent to calling
+   * `this.compose(Polynomial.x + h)`, but is likely to compute the shifted
+   * polynomial much faster.
+   *
+   * @note We *could* do this without a EuclideanRing, by keeping track of the
+   * the multipliers brought down by the derivative, and cancelling as
+   * required. This will require work though, so another time.
+   */
+  def shift(h: C)(implicit ring: Ring[C], eq: Eq[C]): Polynomial[C] = {
+    // The trick here came from this answer:
+    //   http://math.stackexchange.com/questions/694565/polynomial-shift
+    // This is a heavily optimized version of the same idea. This is fairly
+    // critical method to be fast, since it is the most expensive part of the
+    // VAS root isolation algorithm.
+
+    def fromSafeLong(x: SafeLong): C =
+      if (x.isValidInt) {
+        ring.fromInt(x.toInt)
+      } else {
+        val d = ring.fromInt(1 << 30)
+        val mask = (1L << 30) - 1
+
+        @tailrec def loop(k: C, y: SafeLong, acc: C): C =
+          if (y.isValidInt) {
+            k * ring.fromInt(y.toInt) + acc
+          } else {
+            val z = y >> 30
+            val r = ring.fromInt((y & mask).toInt)
+            loop(d * k, z, k * r + acc)
+          }
+
+        loop(ring.one, x, ring.zero)
+      }
+
+    val coeffs: Array[C] = this.coeffsArray.clone()
+    this.foreachNonZero { (deg, c) =>
+      var i = 1
+      var d = deg - 1
+      var m = SafeLong(1L)
+      var k = c
+      while (d >= 0) {
+        m = (m * (d + 1)) /~ i
```

Comment:
Note that we do division, but only on SafeLongs. This is not just for performance, but also required for us to only ask for a `Ring`, rather than a `EuclideanRing`. We always know that `m * (d + 1)` is divisible by `i`, so this is exact.

I should probably add this comment to the code...